File: backend/app/schemas/document.py
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.document import DocumentFamily
from app.schemas.patient import PatientSummary
from app.schemas.types import LocalDatetime


class TemplateOption(BaseModel):
    value: str
    label: str


class TemplateField(BaseModel):
    """Definición de un campo editable de la plantilla.

    `type` gobierna el control que dibuja el cliente: text, textarea, number,
    date, select, boolean o computed. Un campo `computed` no se escribe: su
    valor es la suma de las claves indicadas en `sum` (escalas con puntaje).
    """

    key: str
    label: str
    type: str = "text"
    group: str | None = None
    default: Any = None
    placeholder: str | None = None
    help: str | None = None
    options: list[TemplateOption] | None = None
    sum: list[str] | None = None
    required: bool = False
    wide: bool = False


class TemplateListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    code: str
    version: int
    family: str
    family_label: str
    title: str
    description: str | None
    requires_signature: bool
    study_type: str | None
    field_count: int


class TemplateRead(TemplateListItem):
    fields: list[TemplateField]


class DocumentBase(BaseModel):
    patient_id: int
    encounter_id: int | None = None
    study_id: int | None = None
    practitioner_id: int | None = None
    data: dict[str, Any] = Field(default_factory=dict)

    @field_validator("data")
    @classmethod
    def _clean(cls, value: dict[str, Any]) -> dict[str, Any]:
        """Descarta los campos vacíos: un dato ausente se imprime en blanco."""
        return {
            key: (" ".join(item.split()) if isinstance(item, str) else item)
            for key, item in value.items()
            if item is not None and (not isinstance(item, str) or item.strip())
        }


class DocumentCreate(DocumentBase):
    """Emisión de un documento a partir de una plantilla del catálogo."""

    template_code: str = Field(min_length=2, max_length=32)


class DocumentUpdate(DocumentBase):
    """Corrección de un documento aún en borrador."""


class DocumentVoid(BaseModel):
    reason: str = Field(min_length=4, max_length=255)


class DocumentListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    number: str
    patient_id: int
    patient_name: str
    template_code: str
    template_version: int
    family: str
    family_label: str
    title: str
    status: str
    status_label: str
    practitioner_name: str | None
    study_id: int | None
    encounter_id: int | None
    issued_at: LocalDatetime | None
    created_at: LocalDatetime


class DocumentDetail(DocumentListItem):
    """Documento tal como está guardado, sin la definición de sus campos."""

    template_id: int
    practitioner_id: int | None
    data: dict[str, Any]
    void_reason: str | None
    voided_at: LocalDatetime | None
    is_draft: bool
    is_issued: bool
    patient: PatientSummary


class DocumentRead(DocumentDetail):
    """Detalle más los campos de la versión de plantilla con la que se llenó.

    El formulario del cliente se arma con `fields`, de modo que agregar un
    formato al catálogo no obliga a tocar el frontend.
    """

    fields: list[TemplateField]


class DocumentPreview(BaseModel):
    """Hoja lista para mostrar o imprimir.

    En borrador se rearma con la plantilla vigente; una vez emitido devuelve el
    HTML congelado al momento de la firma.
    """

    id: int
    number: str
    title: str
    status: str
    is_issued: bool
    html: str


class DocumentFamilyOption(BaseModel):
    value: DocumentFamily
    label: str

File: frontend/src/types/document.ts
import type { PatientSummary } from './patient'
import type { StudyType } from './study'

export const DOCUMENT_FAMILIES = ['INFORME', 'FICHA', 'CONSENTIMIENTO', 'ADMINISTRATIVO'] as const
export type DocumentFamily = (typeof DOCUMENT_FAMILIES)[number]

export const DOCUMENT_FAMILY_LABELS: Record<DocumentFamily, string> = {
  INFORME: 'Informe de resultado',
  FICHA: 'Ficha clínica',
  CONSENTIMIENTO: 'Consentimiento y declaración',
  ADMINISTRATIVO: 'Documento administrativo',
}

export const DOCUMENT_STATUSES = ['BORRADOR', 'EMITIDO', 'ANULADO'] as const
export type DocumentStatus = (typeof DOCUMENT_STATUSES)[number]

export const DOCUMENT_STATUS_LABELS: Record<DocumentStatus, string> = {
  BORRADOR: 'Borrador',
  EMITIDO: 'Emitido',
  ANULADO: 'Anulado',
}

/** Controles que sabe dibujar el formulario dinámico. */
const TEMPLATE_FIELD_TYPES = [
  'text',
  'textarea',
  'number',
  'date',
  'select',
  'boolean',
  'computed',
  'odontograma',
] as const
export type TemplateFieldType = (typeof TEMPLATE_FIELD_TYPES)[number]

export interface TemplateOption {
  value: string
  label: string
}

/** Definición de un campo editable, tal como la publica la plantilla. */
export interface TemplateField {
  key: string
  label: string
  type: TemplateFieldType
  group: string | null
  default: string | number | boolean | null
  placeholder: string | null
  help: string | null
  options: TemplateOption[] | null
  /** Sólo en `computed`: claves cuyos valores se suman para obtener el puntaje. */
  sum: string[] | null
  required: boolean
  /** Ocupa el ancho completo de la fila del formulario. */
  wide: boolean
}

export interface DocumentTemplateSummary {
  id: number
  code: string
  version: number
  family: DocumentFamily
  family_label: string
  title: string
  description: string | null
  requires_signature: boolean
  study_type: StudyType | null
  field_count: number
}

/** Datos capturados. Las claves son las de `TemplateField`. */
/** El odontograma guarda un objeto por pieza; el resto de campos, escalares. */
export type DocumentData = Record<string, string | number | boolean | null | object>

export interface DocumentPayload {
  patient_id: number
  encounter_id: number | null
  study_id: number | null
  practitioner_id: number | null
  data: DocumentData
}

export interface DocumentCreatePayload extends DocumentPayload {
  template_code: string
}

export interface DocumentListItem {
  id: number
  number: string
  patient_id: number
  patient_name: string
  template_code: string
  template_version: number
  family: DocumentFamily
  family_label: string
  title: string
  status: DocumentStatus
  status_label: string
  practitioner_name: string | null
  study_id: number | null
  encounter_id: number | null
  issued_at: string | null
  created_at: string
}

export interface ClinicalDocument extends DocumentListItem {
  template_id: number
  practitioner_id: number | null
  data: DocumentData
  void_reason: string | null
  voided_at: string | null
  is_draft: boolean
  is_issued: boolean
  patient: PatientSummary
  /** Campos de la versión con la que se llenó, no de la vigente. */
  fields: TemplateField[]
}

/** Hoja lista para mostrar o imprimir. */
export interface DocumentSheet {
  id: number
  number: string
  title: string
  status: DocumentStatus
  is_issued: boolean
  html: string
}

export interface DocumentFilters {
  search?: string
  patient_id?: number
  family?: string
  status?: string
  template_code?: string
  date_from?: string
  date_to?: string
  page: number
  page_size: number
}
